app/middleware/error-handler.js

```diff
@@ -0,0 +1,47 @@
+'use strict'
+
+const logger = require('../utils/logger')(__filename)
+const { keys } = require('@govuk-pay/pay-js-commons').logging
+const {
+  NotAuthenticatedError,
+  UserAccountDisabledError,
+  NotAuthorisedError,
+  NotFoundError
+} = require('../errors')
+const paths = require('../paths')
+const { CORRELATION_HEADER } = require('../utils/correlation-header')
+const { renderErrorView } = require('../utils/response')
+
+module.exports = function errorHandler (err, req, res, next) {
+  const logContext = {}
+  logContext[keys.CORRELATION_ID] = req.headers[CORRELATION_HEADER]
+
+  if (res.headersSent) {
+    return next(err)
+  }
+
+  if (err instanceof NotAuthenticatedError) {
+    req.session.last_url = req.originalUrl
```

Comment:
Although 99.9/100 the session should be correctly configured for the request at the point of hitting this error handler it might be worth guarding against the possibility of it not, any code that could throw errors here would be handled unpredictably
```js
if (req.session) {
  req.session.last_url = req.originalUrl
}
```

Reply:
good point 👍